### PR TITLE
Map visualization

### PR DESCRIPTION
**Pull request by thorncorona**
# Description

- Added `mapviz.html` map visualization
- Mocked backend using DAO pattern.

# Checklist:

- [x] Code follows the style guidelines
- [x] Code commented in relevant places
- [x] Changes to necessary documentation made
- [x] New and existing unit tests pass locally with changes
- [x] Basic manual testing performed after deploying app

### CODE REVIEW

File: src/main/java/com/google/sps/Mock.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.google.sps;
+
+import com.google.sps.models.PCAPAnalysis;
+
+public class Mock {
```
Comment (arunk054) [inline src/main/java/com/google/sps/Mock.java]: Should be more descriptive than just "Mock" and can you also clarify what "sps" stands for?

File: src/main/java/com/google/sps/Mock.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  public static final String PCAP_ID = "8395f9e1-cc56-45d3-b801-aaa9a2ac8848";
```
Comment (arunk054) [inline src/main/java/com/google/sps/Mock.java]: Leave a comment on what this ID is used for. Is it a unique ID to the PCAP entry in the datastore?

File: src/main/java/com/google/sps/dao/GeolocationDao.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.google.sps.dao;
+
+import com.google.sps.Mock;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class GeolocationDao {
```
Comment (arunk054) [inline src/main/java/com/google/sps/dao/GeolocationDao.java]: This is not really a DAO, you need to abstract it in a interface with getters, and you can extend it with a mock implementation. E.g. https://www.tutorialspoint.com/design_pattern/data_access_object_pattern.htm

File: src/main/java/com/google/sps/dao/GeolocationDao.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  public static Map<String, String> ipCountryMap;
```
Comment (arunk054) [inline src/main/java/com/google/sps/dao/GeolocationDao.java]: Avoid having static functions and entries, that is generally not a good practice.

File: src/main/java/com/google/sps/models/PCAPAnalysis.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.google.sps.models;
+
+public class PCAPAnalysis {
```
Comment (arunk054) [inline src/main/java/com/google/sps/models/PCAPAnalysis.java]: if you can make this an interface it will be more extensible

File: src/main/java/com/google/sps/models/PCAPAnalysis.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  public PCAPAnalysis(String id, String[] ips) {
+    this.id = id;
+    this.ips = ips;
+  }
+
+  public String getId() {
+    return id;
+  }
+
+  public void setId(String id) {
```
Comment (arunk054) [inline src/main/java/com/google/sps/models/PCAPAnalysis.java]: you should not have a setter for ID, as it should remain unchanged after object creation.

File: src/main/java/com/google/sps/models/PCAPAnalysis.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package com.google.sps.models;
+
+public class PCAPAnalysis {
+
+  private String id;
+  private String[] ips;
+
+  public PCAPAnalysis(String id, String[] ips) {
```
Comment (arunk054) [inline src/main/java/com/google/sps/models/PCAPAnalysis.java]: Can you add a comment on what is the purpose of this class. What is it trying to achieve by encapsulating a id and set of IPs.

File: src/main/java/com/google/sps/Mock.java (hunk 1 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/Mock.java]: Don't know what it stands for haha. It was the name of our package in the starter project. Changed package name

I changed it to `MockData`.

File: src/main/java/com/google/sps/Mock.java (hunk 2 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/Mock.java]: Just random data

File: src/main/java/com/google/sps/dao/GeolocationDao.java (hunk 4 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/dao/GeolocationDao.java]: Fixed, integrated into DAO mock

File: src/main/java/com/google/sps/models/PCAPAnalysis.java (hunk 7 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/models/PCAPAnalysis.java]: Just a placeholder until I integrate w/ Eric.

File: src/main/java/com/google/sps/models/PCAPAnalysis.java (hunk 6 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/models/PCAPAnalysis.java]: Fixed

File: src/main/java/com/google/sps/models/PCAPAnalysis.java (hunk 7 above)
Comment (arunk054) [reply src/main/java/com/google/sps/models/PCAPAnalysis.java]: In that case add a comment and/or TODO saying it should be modified / removed after some integration.

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/ipgeolocation")
+public class IPGeolocationServlet extends HttpServlet {
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String id = getParameter(request, "PCAPId", "");
```
Comment (arunk054) [inline src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: You could create a default PCAPID and store it as a constant. Your getParameter could return that instead. For now add a TODO may b until you add a default pcap id to your data store / mocks.

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+      if (analysis == null) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      Map<String, Integer> countryCount = new HashMap<>();
+
+      for (String ip: analysis.getIps()) {
+        String country = GeolocationDao.getCountryByIP(ip);
```
Comment (arunk054) [inline src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: Are you taking care of case sensitive names in the DAO I believe. This assumes that the caller is sure the returned value is always case agnostic. Add a comment for that assumption.

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        countryCount.put(country, countryCount.getOrDefault(country, 1));
+      }
+
+      // mock
+      countryCount.put("Armenia", 43);
+      countryCount.put("Jordan", 67);
+      countryCount.put("Estonia", 99);
+      countryCount.put("South Korea", 21);
+      countryCount.put("Canada", 66);
```
Comment (arunk054) [inline src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: I believe all of this is going to be in your Mock DAO?

File: src/main/java/com/google/sps/models/PCAPAnalysis.java (hunk 7 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/models/PCAPAnalysis.java]: Changed.

File: src/main/java/com/google/sps/models/PCAPAnalysis.java (hunk 5 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/models/PCAPAnalysis.java]: Fixed.

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/ipgeolocation")
+public class IPGeolocationServlet extends HttpServlet {
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String id = getParameter(request, "PCAPId", "");
+      PCAPAnalysis analysis = AnalysisDao.getAnalysis(id);
```
Comment (arunk054) [inline src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: You need to instantiate a DAO object in your init() method. That DAO instance can be a mock or can be talking to your data store or something else, based on the servlet config. Then in your servlet method here you will always call analysisDaoInstance.getAnalysis(id) irrespective of mocks or real. That will make it easier to inject mocks.

**Comment by thorncorona** [comment]
@arunk054 new changes pushed up!

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java (hunk 10 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: Removed.

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java (hunk 11 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: In new commits

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java (hunk 9 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: In new commits!

File: src/main/java/com/google/sps/servlets/IPGeolocationServlet.java (hunk 8 above)
Comment (thorncorona) [reply src/main/java/com/google/sps/servlets/IPGeolocationServlet.java]: Yep I changed the implementation in the new commits. For now I just set a default mock pcap

File: scripts/BenchmarkGeoDB.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
+import com.maxmind.geoip2.DatabaseReader;
+import com.maxmind.geoip2.exception.AddressNotFoundException;
+import com.maxmind.geoip2.model.CityResponse;
+import java.io.File;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+public class BenchmarkGeoDB {
```
Comment (arunk054) [inline scripts/BenchmarkGeoDB.java]: submit your benchmark stuff as a separate PR so it is easier for Philip to see.

File: src/main/java/com/google/netpcapanalysis/interfaces/models/PCAP.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package com.google.netpcapanalysis.interfaces.models;
+
+import java.net.InetAddress;
+import java.util.*;
+
+public interface PCAP {
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/interfaces/models/PCAP.java]: PCAP has more to it than IP addresses. You should choose a different name for the interface. Melba has a PCAPData class that contains all the n tuples in a record, so a PCAP file is a list of PCAPData objects. So basically you should try to use that if you can, by creating a list of PCAP objects. You might only use the IP field that is fine and your mock would only mock IP.

File: src/main/java/com/google/netpcapanalysis/models/MockPCAP.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  public String getId() {
+    return id;
+  }
+
+  @Override
+  public List<InetAddress> getIPs() {
+    return ips;
+  }
+
+  public static InetAddress genIP() {
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/models/MockPCAP.java]: no reason this should be static. It can be a non-static method.

File: src/main/java/com/google/netpcapanalysis/models/MockPCAP.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        ip.isAnyLocalAddress() ||
+        ip.isLinkLocalAddress() ||
+        ip.isSiteLocalAddress()) {
+      try {
+        ip = InetAddress.getByName(
+            r.nextInt(256) + "." +
+                r.nextInt(256) + "." +
+                r.nextInt(256) + "." +
+                r.nextInt(256)
+        );
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/models/MockPCAP.java]: Did you do the rough math or a statistical computation on roughly how many iterations it takes before you get the IP you need. Seems a bit overkill for just mock data. You could as well run this offline in a separate script and then copy a set of valid IPs to a text file.

File: src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+import java.net.URL;
+
+public class GeolocationDaoImpl implements GeolocationDao {
+
+  private File database;
+  private DatabaseReader reader;
+
+  public GeolocationDaoImpl() {
+    try {
+      URL geoDBUrl = GeolocationDaoImpl.class.getClassLoader().getResource("GeoLite2-City.mmdb");
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java]: Declare the resource file name as a final static variable. Ideally we could read this from a properties file but that is future.

**Review by arunk054 — COMMENTED** [review]
Looks much better. I think we need to work on abstracting the PCAP object out a little bit, coordinating with Melba's changes.

File: src/main/java/com/google/netpcapanalysis/models/MockPCAP.java (hunk 15 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/models/MockPCAP.java]: I just copied it from the benchmark script. The time is pretty negligible. I used it to generate IPs for the benchmark script and 400K addresses didn't make the script much slower.

File: src/main/java/com/google/netpcapanalysis/models/MockPCAP.java (hunk 14 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/models/MockPCAP.java]: Fixed.

File: src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java (hunk 16 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java]: Fixed

File: scripts/BenchmarkGeoDB.java (hunk 12 above)
Comment (thorncorona) [reply scripts/BenchmarkGeoDB.java]: Removed from branch

File: src/main/java/com/google/netpcapanalysis/interfaces/models/PCAP.java (hunk 13 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/interfaces/models/PCAP.java]: Can I use PCAP as a stub, and let Melba merge on top so we can use hers?

Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/interfaces/models/PCAP.java]: @mnuzen Are you okay with merging your class into the interface?

**Comment by thorncorona** [comment]
Rebased.

**Comment by thorncorona** [comment]
@arunk054 I merged the PCAP models w/ Erik/Melba, can I get approved?

File: README.md [hunk 17]
```diff
@@ -4,3 +4,12 @@ This repo contains our pod's 2020 STEP Capstone project.
 
 Authors: handeland@google.com, jevingu@google.com, mnuzen@google.com 
 Reviewers: arunkaly@, promanov@
+
+
+## To use geolocation: 
+
+1. Download `GeoLite2-City.mmdb` from GDrive
+2. Put in repo under `data/GeoLite2-City.mmdb`
```
Comment (arunk054) [inline README.md]: Try to use the .git folder for this or some other folder that you can mark in `.gitignore` so you don't accidentally commit this file.

File: src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+import com.maxmind.geoip2.model.CityResponse;
+import com.maxmind.geoip2.record.Country;
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URL;
+
+public class GeolocationDaoImpl implements GeolocationDao {
+
+  private static final String GEO_DB_LOCATION = "GeoLite2-City.mmdb";
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java:14]: Is this not under `data/` according to your README?

File: src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+  private File database;
+  private DatabaseReader reader;
+
+  public GeolocationDaoImpl() {
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java:18]: You could also have an overloaded constructor to pass in the GEO DB file path so that would make it a bit more flexible.

File: src/main/java/com/google/netpcapanalysis/mockdata/MockDataLoader.java [hunk 20]
```diff
@@ -1,33 +1,31 @@
...
   private String csvFile = "data.csv"; // CSV located in project dir /webapp
 
-  public MockDataLoader(){
+  public MockDataLoader() {
 
   }
 
   public void LoadData(ArrayList<PCAPdata> dataTable) { // uploads to datastore
-    GenericPCAPDao dataBase = new GenericPCAPDaoImpl();
+    PCAPDao dataBase = new PCAPDaoImpl();
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/mockdata/MockDataLoader.java:24]: You are creating the DAO every time instead of using a factory that maintains a global instance. So could lead to problems if two people trying to call this function on your website. The PCAPDaoImpl should be modified to a factory pattern like getInstance() instead allowing a constructor. May be do it on your next iteration.

File: src/main/tests/models/PCAPTest.java [hunk 21]
```diff
@@ -0,0 +1,5 @@
+package models;
+
+public class PCAPTest {
```
Comment (arunk054) [inline src/main/tests/models/PCAPTest.java:3]: what is this class for btw?

**Review by arunk054 — APPROVED** [review]

File: README.md (hunk 17 above)
Comment (thorncorona) [reply README.md]: Yep in gitignore

File: src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java (hunk 18 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java:14]: It's mapped through Java resources.

File: src/main/tests/models/PCAPTest.java (hunk 21 above)
Comment (thorncorona) [reply src/main/tests/models/PCAPTest.java:3]: Scaffolding for Melba/Erik's stuff I think.

File: src/main/java/com/google/netpcapanalysis/mockdata/MockDataLoader.java (hunk 20 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/mockdata/MockDataLoader.java:24]: SG
